### PR TITLE
[SofaPython3] Add a LinkPath object in both the binding and plugin.

### PR DESCRIPTION
**Pull request by damienmarchal**
This PR is following the discussion in #197 

The existing syntax to make links in SofaPython3 is the following one:
```python
node.addObject("MeshObjLoader", name="loader")
node.addObject("MechanicalObject", position=node.loader.position.getLinkPath())
```

In addition to be very verbose the getLinkPath() is returning a string (eg: "@/node/loader.position").
Representing linkpath with string has the drawback of forcing the two objects to be part of the same simulation graph which is not always the case.

Example of code that does not work as expected:
```python
def MyPrefab(target):
	node = Sofa.Core.Node("MyPrefab")
	node.addObject("MeshObjLoader", name="loader")
	node2.addObject("MechanicalObject", position=target)
	return node

def createScene(root):
	root.addObject("MeshObjLoader", name="loader")
	root.addChild( MyPrefab(loader.position.getLinkPath()) )
```

It does not work because string returned by getLinkPath() is a path query into the scene-graph to locate the object or the data
field to attach not into the one of the graph the destination object is. 

To solve this issue this PR introduce a LinkPath object which is dedicated to point to a sofa object.
This LinkPath object is exposed in python and can be used to make parenting between objects and data without the need of graph to path then path to graph conversion (so it is also much faster.. but we don't really care here right ?).

For consistency with the "value" in data field the PR expose the LinkPath with a dedicated "linkpath" attribute.
So the code is now the following:
```python
root.addObject("MeshObjLoader", name="loader")
root.addObject("MechanicalObject", name="p1", position=root.loader.position.value)      # Copy the value
root.addObject("MechanicalObject", name="p2", position=root.loader.position.linkpath)   # Make a link between data without the need for string based query

root.addObject("Mapping", name="loader", input=p1.linkpath)
```

### CODE REVIEW

**Comment by damienmarchal** [comment]
This PR really deserve some attention :)

File: bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp [hunk 1]
```diff
@@ -185,6 +188,26 @@ py::object getObject(Node &n, const std::string &name, const py::kwargs& kwargs)
     return py::none();
 }
 
+void setFieldsFromPythonValues(Base* self, const py::kwargs& dict)
+{
+    // For each argument of the addObject functin we check if this is a
```
Comment (alxbilger) [inline bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp:193]:
```suggestion
    // For each argument of the addObject function we check if this is a
```
I think the sentence is not finished

File: bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp [hunk 2]
```diff
@@ -185,6 +188,26 @@ py::object getObject(Node &n, const std::string &name, const py::kwargs& kwargs)
...
+        {
+            auto* data = self->findData(py::str(kv.first));
+            if(data)
+                BindingBase::SetData(data, py::cast<py::object>(kv.second));
+
+            auto* link = self->findLink(py::str(kv.first));
+            if(link)
+                BindingBase::SetLink(link, py::cast<py::object>(kv.second));
+        }
+    }
```
Comment (alxbilger) [inline bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp:208]:
```suggestion
    for(auto [key, value] : dict)
    {
        if(py::isinstance<LinkPath>(value))
        {
            auto* data = self->findData(py::str(key));
            if(data)
                BindingBase::SetData(data, py::cast<py::object>(value));

            auto* link = self->findLink(py::str(key));
            if(link)
                BindingBase::SetLink(link, py::cast<py::object>(value));
        }
    }
```

This suggestion using a structured binding is not mandatory, but I think it's a case where it helps the reading

**Review by alxbilger — CHANGES_REQUESTED** [review]

File: Plugin/src/SofaPython3/LinkPath.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{
+    targetBase = target;
+    targetData = nullptr;
+}
+
+LinkPath::LinkPath(sofa::core::objectmodel::BaseData* target)
+{
+    if(target->getOwner())
+    {
+        targetBase = target->getOwner();
```
[the remark this answers is not in the document]
Comment (damienmarchal) [reply Plugin/src/SofaPython3/LinkPath.cpp]: Thank for pointing this. The condition is not needed.

File: bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp (hunk 2 above)
Comment (alxbilger) [reply bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp:208]: I don't see this as resolved

**Comment by damienmarchal** [comment]
If we agree this is the new API yes it is very easy to deprecate getLinkPath...but this will emit a lot of warning in every python scene. So I recommend deprecating the getLinkPath() thing for 22.12

File: bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp (hunk 1 above)
Comment (damienmarchal) [reply bindings/Sofa/src/SofaPython3/Sofa/Core/Binding_Node.cpp:193]: DIdn't had commited the suggestion ?

**Comment by damienmarchal** [comment]
@hugtalbot so what ?

**Comment by damienmarchal** [comment]
This PR is adding a new feature linkpath that is a "non string" approach to make link between data or objects. 
The PR was proposed to implement the conclusion of https://github.com/sofa-framework/SofaPython3/issues/197 
The issue 197 wanted being explicit 
```python
node.addObject("toto", position=otherObject.position.linkpath)  # makes a link without path conversion explicitely
node.addObject("toto", position=otherObject.position.value)  # makes a copy to initialize
```
instead of: 
```python
node.addObject("toto", position=otherObject.position)  # makes a link without path conversion implicitely
node.addObject("toto", position=otherObject.position.value)  # makes a copy to initialize
```

Hugo suggested to deprecate getLinkPath()... I'm not sure it is the proper PR to do that because it is not wise to deprecate a long standing feature by a modernized replacement. 

But I see two options:
- JérémieA pointed that in InSimo's fork they are using "path" in place of "getLinkPath()". It may be wise to replace getLinkPath()
In that case  "myobject.linkpath" and "myobject.path" are not returning the same thing. And modernizing the examples are:

```python
# The existing 
 # makes a link with a stringyfied path (so generation and parse) and to work this imply being in the same graph for path resolution suceed at init and so on). 
node.addObject("toto", position=position.getLinkPaht()) 

# This version (from InSimo) is strictly equivalent to getLinkPath (same drawbacks) except that it use a python attribute to make the syntax more "slick"... we may want to go to such a thing  
node.addObject("toto", position=position.path)  

# This version (from the PR) is *not* a replacement of getLinkPath it is superset. 
node.addObject("toto", position=position.linkpath)  # makes a link without path conversion
print("This is a to link object", position.linkpath) 
node.addObject("toto", position=str(position.linkpath))  # makes a link with path conversion
print("This is the link representation as a path", str(position.linkpath)) 
```

**Comment by damienmarchal** [comment]
Hi @hugtalbot, thank for pointing I will look at that after the release.

**Pull request merged**